process/common.go

```diff
@@ -867,6 +875,11 @@ func UnmarshalShardHeader(marshalizer marshal.Marshalizer, hdrBuff []byte) (data
 		return hdr, nil
 	}
 
+	hdr, err = UnmarshalHeaderWithValidatorStats(marshalizer, hdrBuff)
```

Comment:
why do we need a code like this ? can't we have a factory/interface for this unmarshal and 3 implementations. Why should we try unmarshalling for header v2 and only after doing for the one with validators stats ? It is double work on CPU. In case of importDB, it is triple work.

Reply:
Unfortunately could not be resolved for all the cases, as on header interceptor you will receive some bytes. All you know is that this should be a header. Then you should try to unmarshall in all these variants. To optimize this stuff at least we put the most probable implementation header, like v2 for now, as the first try.

Reply:
can't we have sovereignHeaderInterceptor and hearderV2Interceptor ? I think this part has to be refactored.

Reply:
I can get the task of injecting an abstract factory that will get only the marshaled bytes and try to output a headerHandler. We will have 2 versions: one for normal chain operation, and one for the sovereign. Then we can try to optimize it further if we want. This will be a whole new PR.